Add support for cluster autoscaling

tuommaki opened the pull request:
In order to support cluster autoscaling, there must be reconciliation of
worker ASG min and max values and synchronization of present desired
configuration (to status field).

To migrate older clusters, scaling min/max values are taken from present
worker count.

Towards https://github.com/giantswarm/giantswarm/issues/4543

tuommaki commented:
**NOTE:** This is currently still WIP and hence CI is also failing.

MarcelMue commented on service/controller/v21/adapter/guest_auto_scaling_group.go:
```diff
@@ -22,9 +23,19 @@ type GuestAutoScalingGroupAdapter struct {
+	}
+
+	if maxWorkers == minWorkers {
+		maxWorkers++
```
Why are we doing this here?

tuommaki replied:
Because in previous version (little bit down here below) it used to be like this:
```go
	a.ASGMaxSize = workers + 1
	a.ASGMinSize = workers
```

My guess for why this is like this is that this gives room for cluster movement when rolling nodes for any reason (although upgrade is handled still little bit differently), but I might be wrong. Nevertheless, that's how it used to be.

teemow replied:
Might be legacy from before we started using cloudformations.

tuommaki replied:
No, it's actually needed and we discussed about this. I also added comment here, but this source code snippet in discussion view is old.

teemow replied:
The upgrade first starts new instances. Once they are up and running we start with the draining. On bigger clusters we definitely start bigger batches than a single node. 

I just started a test cluster with 12 nodes on AWS. Max was set to 13. The rolling update of the ASG started 4 new workers.

Are you really sure Max has any effect on rolling updates?

tuommaki replied:
Settled with max == min for now. After discussions this should be correct nowadays with lifecycle hooks etc.. Testing will prove this a bit later.

tuommaki commented:
Code to update `Status.Cluster.Scaling.DesiredCapacity` from worker ASG is missing here and I'll provide it as a separate pull request in order to keep this one smaller.

xh3b4sd commented on service/controller/v21/adapter/guest_auto_scaling_group.go:
```diff
@@ -61,3 +76,15 @@ func workerCountRatio(workers int, ratio float32) string {
 	return strconv.Itoa(rounded)
 }
+
+func minDesiredWorkers(minWorkers, maxWorkers, statusDesiredCapacity int) int {
```
I don't follow here. Why is the value not just always `minWorkers`? Can we explain this in the godoc with some examples?

tuommaki replied:
Added comment. Maybe that helps?

xh3b4sd commented on service/controller/v21/adapter/guest_outputs.go:
```diff
@@ -23,11 +23,13 @@ func (a *GuestOutputsAdapter) Adapt(config Config) error {
 	a.Worker.InstanceType = config.StackState.WorkerInstanceType
-	a.Worker.CloudConfig.Version = config.StackState.WorkerCloudConfigVersion
+	a.Worker.Max = config.StackState.WorkerMax
+	a.Worker.Min = config.StackState.WorkerMin
```
What is the reason to track this in the outputs? During stack state transitioning the output values might not be accessible IIRC. Getting the values from the ASG when we need it would be more reliable as the occasionally unavailable redirect here, wouldn't it?

tuommaki replied:
Stackstate transitioning is the mechanism involved also in scaling vs. update decisions, but you have a good point. I'll do some investigation and see if we could get rid of this. Getting this right was slightly difficult and I now have even a bug in there (due to nil value for `WorkerCount`) so I'll take a look at this.

tuommaki replied:
Ok. I'm working on this. I'm pretty sure I can get rid of tracking this via CloudFormation stack output, but I keep the state in `StackState` still, it's just read directly from ASG instead of stack output. This way I can also read the present `Desired` property at the right time and later just save it in Status field in `ApplyUpdate()` -> no need for additional resource then. I'm still in the middle of finalizing this, but so far it looks good. Thanks for your comment here! :heart:

xh3b4sd replied:
Staying in the same resource makes sense when the lifecycle of information and behaviour is actually aligned.

xh3b4sd commented on service/controller/v21/resource/cloudformation/spec.go:
```diff
@@ -32,12 +32,14 @@ type StackState struct {
 	ShouldScale  bool
 	ShouldUpdate bool
 
+	WorkerCloudConfigVersion string
```
I appreciate your boyscouting and consistency here. <3 :D

tuommaki commented:
Allright! I removed CF stack output magic regarding worker count and now fetch ASG state in `resource/cloudformation.GetCurrentState()` and store it in `StackState` and later in `ApplyUpdateChange()` I write the `DesiredCapacity` status field out.

Took first smoke out on `ginger` and this looks good:
• Spinned up a cluster with 3 workers (i.e. no specification for CR so `migration` resource took place and set min: 3, max: 3)
• Cluster came up fine
• I changed max 3 -> 6 in CR spec
• ASG change applied, nodes stayed at 3
• I changed Desired to 5 in AWS Console -> 5 nodes to cluster
• CR status got up to date once reconciliation loop executed
• I changed spec to min: 3, max: 4 -> now changes applied and current desired & max: 4 (and nodes downscaled)

kopiczko commented on service/controller/v21/resource/cloudformation/current.go:
```diff
@@ -217,22 +216,53 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
+			if len(o.AutoScalingGroups) == 0 {
+				return StackState{}, microerror.Maskf(notExistsError, "asg for name %s", workerASGName)
+			}
+			asg := o.AutoScalingGroups[0]
```
Should we handle case when `len(o.AutoSaclingGroups) > 1` or `!= 1` in general?

tuommaki replied:
Not yet. Time for that will come when we start working on node pools, but that will change many places so radically any way that I don't consider changes related to that right now. There's only one worker ASG and it's selected by ARN so length of results is either 0 or 1.

kopiczko commented on service/controller/v21/resource/cloudformation/current.go:
```diff
@@ -217,22 +216,53 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
+				return StackState{}, microerror.Mask(err)
+			}
+			if len(o.AutoScalingGroups) == 0 {
+				return StackState{}, microerror.Maskf(notExistsError, "asg for name %s", workerASGName)
```
We usually have `notExistsError`. But isn't it isn't handled anywhere I'd replace them with `ececutionFailedError` to make that notion clear.

tuommaki replied:
Ack

kopiczko commented on service/controller/v21/resource/cloudformation/update.go:
```diff
@@ -27,6 +28,25 @@ func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange inte
+		_, err = r.g8sClient.ProviderV1alpha1().AWSConfigs(newObj.Namespace).UpdateStatus(newObj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
```
Should the reconciliation be cancelled here? New update event is about to arrive.

tuommaki replied:
I'm not actually 100% sure about this. When I was writing this my feeling was that no matter what you do here, it would get into infinite loop, but that's not the case. When cancelling reconciliation and working on updated version, wouldn't it end up to this same state? I kind of stopped worrying about this after I saw in practice that this was working correctly as-is (meaning it updates status according to refresh period of 5min).

kopiczko replied:
I would make sense to me if `UpdateStatus` wouldn't generate an event, but I thought it isn't the case.

xh3b4sd replied:
This is an ongoing issue I cannot wrap my head around. Like, how should the reconciliation work in a perfect world? You maybe only want to react on changes to the metadata, spec, or status. All other triggers can be ignored in some cases. So the framework should maybe support something like this, which is probably a bit difficult to implement and maybe not even the right place in the framework. I am wondering if upstream is interested in such feature.

xh3b4sd replied:
And at the end we probably want to have separate resources to reflect the separate logic steps. We want to break down the CF resource anyway.

kopiczko replied:
> Like, how should the reconciliation work in a perfect world? You maybe only want to react on changes to the metadata, spec, or status.

Let's narrow the scope to events generated by updating `status` part of the object which is the case here. Should a resource always cancel reconciliation after doing so? I personally think the resource shouldn't update status more than once per reconciliation loop and yes it should always cancel reconciliation after updating status. Is there any scenario you see where that would be disruptive?

xh3b4sd replied:
I have nothing in mind concerning disruptions. Agree on cancelation.

tuommaki commented:
> For my understanding. Is this code purpose to reflect changes made by the autoscaler in the CF stacks?

Forgot to answer this. Yes, this is to enable running of `cluster-autoscaler` in clusters with this release out-of-the-box.

xh3b4sd commented:
We check more on e2e tests in the next steps.
